sample_algorithms/preemptive.py

```diff
@@ -0,0 +1,270 @@
+      if 'position' in obj and distance <= cameraSensorRadius and abs(direction - car['direction']) <= cameraSensorFOVAngle / 2 :
+        cameraData.append(obj)
+  if len(cameraData) > 0:
+    print(cameraData)
+  return cameraData
+
+def addToState(journeys, state):
+  for journey in journeys:
+    start = {"geometry": {"type": "Point", "coordinates": [journey['origin']['lng'], journey['origin']['lat']]}, "type": "Feature", "properties": {}}
+    end = {"geometry": {"type": "Point", "coordinates": [journey['destination']['lng'], journey['destination']['lat']]}, "type": "Feature", "properties": {}}
+    newRoute = R.getRoute(INP_FILE, start, end)['path']
+    preprocess(newRoute)
+    state.append(createNewCar(len(state), journey['_id'], baseRoute=newRoute))
+    print('.')
+
+def get_distance(start, end):
+  lat1 = math.radians(start[1])
+  lng1 = math.radians(start[0])
+  lat2 = math.radians(end[1])
+  lng2 = math.radians(end[0])
+  d_lat = lat2-lat1
+  d_lng = lng2-lng1
+  a = math.sin(d_lat/2)**2 + math.cos(lat1)*math.cos(lat2)*math.sin(d_lng/2)**2
+  c = 2*math.atan2(math.sqrt(a), math.sqrt(1-a))
+  R = 6371e3
+
+  return R*c
+
+def get_direction(start, end):
+  lat1 = math.radians(start[1])
+  lng1 = math.radians(start[0])
+  lat2 = math.radians(end[1])
+  lng2 = math.radians(end[0])
+  d_lng = lng2-lng1
+  y = math.sin(d_lng)*math.cos(lat2)
+  x = math.cos(lat1)*math.sin(lat2)-math.sin(lat1)*math.cos(lat2)*math.cos(d_lng)
+  return (math.degrees(math.atan2(y, x))+270)%360
+
+def preprocess(route):
+  for i in range(len(route)-1):
+    start = route[i]
+    end = route[i+1]
+
+    props = R.getProperties(INP_FILE, start, end)
+    maxSpeed_km_h = MAX_SPEED_KM_H
+    if 'maxspeed' in props:
+      try:
+        maxSpeed_km_h = int(props['maxspeed']) #TODO: Will break with mph or any suffix
+      except Exception as err:
+        maxSpeed_km_h = MAX_SPEED_KM_H
+        print(err)
+
+    dist = get_distance(start, end)
+    time = dist/(maxSpeed_km_h*1000/3600)
+    end.append({'timeLeft': time, 'totalTime': time, 'maxSpeed': maxSpeed_km_h})
+
+def add(v1, v2):
+  return [v1[0]+v2[0], v1[1]+v2[1]]
+
+def sub(v1, v2):
+  return [v1[0]-v2[0], v1[1]-v2[1]]
+
+def scale(v, s):
+  return [s*v[0], s*v[1]]
+
+def scheduleNewRoute(car):
+  start = car['baseRoute'][0]
+  if (isNodeLocked(start)):
+    return False
+  car['route'] = deepcopy(car['baseRoute'])
+  car['position'] = car['baseRoute'][0]
+  car['direction'] = get_direction(car['route'][0], car['route'][1])
+  car['lockedNode'] = start
+  return True
+
+def isEqualNodes(node1, node2):
+  return node1[0] == node2[0] and node1[1] == node2[1]
+
+def isNodeLocked(node):
+  for car in state:
+    if isEqualNodes(car['lockedNode'], node):
+      return True
+  return False
+
+def switchNodeLock(car, start, end):
+  if (isEqualNodes(car['lockedNode'], start)):
+    if (isNodeLocked(end)):
+      return False
+
+    car['lockedNode'] = end
+  return True
+
+def executeLocalAlgorithm(car):
```

Comment:
what is the point of this function? I think moveCar is more readable than executeLocalAlgorithm